lib/agents/gossip-agent.js

```diff
@@ -42,18 +43,48 @@ module.exports = class GossipAgent extends ContinuityAgent {
   }
 
   async _gw({peer}) {
-    const {ledgerNodeId} = this;
+    const {ledgerNode, ledgerNodeId} = this;
     let result;
     let err;
     do {
       try {
         result = await _gossip.gossipWith(
-          {callerId: this.creatorId, ledgerNode: this.ledgerNode, peer});
+          {callerId: this.creatorId, ledgerNode, peer});
       } catch(e) {
         err = e;
         // if there is an error with one peer, do not stop cycle
         logger.debug('non-critical error in gossip', {err, peer});
       }
+      // process events acquired from peer
+      if(result && result.events) {
+        const {events, needed} = result;
+        try {
+          // hash events and validate ancestry of all the events in the DAG
+          const {eventMap} = await _events.prepBatch(
+            {events, ledgerNode, needed});
+          let deferredEvents = events;
+          let processedEventHashes;
+          // validate events and operations using ledger validators deferring
+          // processing of events until the local node has generated the
+          // blockHeight required by the event's basisBlockHeight
+          do {
+            const {eventBlock: {block: {blockHeight}}} =
+              await ledgerNode.blocks.getLatestSummary();
+            ({deferredEvents, processedEventHashes} = await _events.addBatch({
+              blockHeight, events: deferredEvents, eventMap, ledgerNode,
+              processedEventHashes
+            }));
+            if(deferredEvents.length > 0) {
+              // if events were deferred, wait for a new block and retry
+              await delay(250);
```

Comment:
If we were to have a listener for block creation, it would go here, but I'm not sure there's much benefit to that complexity.  We expect that a short time after `addBatch` on L73 completes one or more new blocks will be created.  I believe it's possible that an add block event may have already fired by this point so we would likely have to do some promiseHelperLib.race() thing waiting on listener or some hard timeout anyway.

Reply:
Ok, we might just want to leave a comment about considering doing some kind of push notification in the future but that there might be race conditions that make it complicated. Nothing to do right now.